pubsub/google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py

```diff
@@ -125,17 +125,31 @@ def drop(self, items):
         Args:
             items(Sequence[DropRequest]): The items to drop.
         """
-        self._manager.leaser.remove(items)
-        self._manager.maybe_resume_consumer()
+        # If the manager is in the process of being shut down, the leaser might
+        # not exist on it anymore, thus we need to obtain our own reference to
+        # it and check it for None to avoid errors.
+        # Nevertheless, calling maybe_resume_consumer() in the manager shutdown
+        # state is still fine, as it is effectively a no-op at that point.
+        leaser = getattr(self._manager, "leaser", None)
+        if leaser is not None:
+            leaser.remove(items)
+            self._manager.maybe_resume_consumer()
```

Comment:
This seems racy:  should it be protected by the `_closing` lock on the manager?

Reply:
You mean the `maybe_resume_consumer()` method or the entire `dispatcher.drop()` method?

It is _likely_ to be fine based on the logs and interaction between the threads, however, "likely" is both not a strong enough word, and it's too dependent on the actual implementation. Having an explicit lock wold make it easier to reason about the code without relying on implicit assumptions, will add it.

(as for the leaser add() / remove() methods, that [other PR](https://github.com/googleapis/google-cloud-python/pull/7948/commits/9b010605f0796a73f3a0e65bee4cba9c6727238e) handles that, will respond there tomorrow).